feat(ast_tools): support `#[scope(exit_before)]`

DonIsaac opened the pull request:
Closes #6311.

Adds support for `#[scope(exit_before)]`, which is the opposite of `#[scope(enter_before)]`

DonIsaac commented:
* **#6350** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6350?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @DonIsaac and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

Dunqing commented:
From the description of #6311, we should be supporting `#[scope(exit_before)]` rather than `#[scope(exit_after]`. Let me know if I got it wrong.

overlookmotel reviewed (changes requested):
As mentioned in #6311, I think `#[scope(exit_before)]` is better than `#[scope(exit_after)]`, because then when you read the type def, the `#[scope(exit_before)]` attr is at the point where the scope is exited.

e.g.:

```rs
pub struct TSConditionalType<'a> {
    pub span: Span,
    pub check_type: TSType<'a>,
    #[scope(enter_before)]
    pub extends_type: TSType<'a>,
    pub true_type: TSType<'a>,
    #[scope(exit_before)]
    pub false_type: TSType<'a>,
    pub scope_id: Cell<Option<ScopeId>>,
}
```

The fields that are inside the scope are `extends_type` and `true_type`, and they are bounded by the `#[scope(...)]` attrs on each side.

Is there a reason why you decided against that?

overlookmotel reviewed (dismissed):
`Traverse` codegen also needs to support the new exit attr.

https://github.com/oxc-project/oxc/tree/main/crates/oxc_traverse/scripts

You can see in #6351 that `walk.rs` is not exiting the scope when it should: https://github.com/oxc-project/oxc/pull/6351/files#diff-f70f8cd02df83ed234c9b27dcb7ea2c948b2ab6110952e87eadaf0ddafbd6b9a

Boshen commented on tasks/ast_tools/src/markers.rs:
```diff
@@ -64,7 +64,10 @@ pub struct VisitMarkers {
 /// A struct representing `#[scope(...)]` markers
 #[derive(Default, Debug)]
 pub struct ScopeMarkers {
+    /// `#[scope(enter_before)]`
```
Unintentional commented out code?

overlookmotel replied:
I believe this is intentional. It's a doc comment explaining what attribute `ScopeMarkers::enter_before` represents.

DonIsaac commented:
> `Traverse` codegen also needs to support the new exit attr.
> 
> https://github.com/oxc-project/oxc/tree/main/crates/oxc_traverse/scripts
> 
> You can see in #6351 that `walk.rs` is not exiting the scope when it should: https://github.com/oxc-project/oxc/pull/6351/files#diff-f70f8cd02df83ed234c9b27dcb7ea2c948b2ab6110952e87eadaf0ddafbd6b9a

Fixed. Ready for another review.

overlookmotel commented on crates/oxc_traverse/scripts/lib/walk.mjs:
```diff
@@ -92,6 +94,20 @@ function generateWalkForStruct(type, types) {
+          `in '${type.name}', but that field is not visited`,
+      );
+    }
+    // else {
+    //   exitScopeField = visitedFields[visitedFields.length - 1];
+    // }
```
Remove commented-out code.

```suggestion
```

overlookmotel commented on crates/oxc_traverse/scripts/lib/walk.mjs:
```diff
@@ -75,7 +75,9 @@ function generateWalkForStruct(type, types) {
-  let scopeEnterField;
+  let scopeEnterField,
+    /** @type {Field | undefined} */
+    exitScopeField;
```
Could you name this var `scopeExitField` for consistency with `scopeEnterField`?

overlookmotel commented on tasks/ast_tools/src/generators/visit.rs:
```diff
@@ -525,6 +527,18 @@ impl<'a> VisitBuilder<'a> {
+                        exit_scope_at.is_none(),
+                        "Scopes cannot be exited more than once. Remove the extra `#[scope(exit_before)]` attribute(s)."
+                    );
+                    let scope_leave = &scope_events.1;
+                    result = quote! {
+                        #scope_leave
+                        #result
+                    };
```
Rename var for consistency.

```suggestion
                    let scope_exit = &scope_events.1;
                    result = quote! {
                        #scope_exit
                        #result
                    };
```

overlookmotel reviewed (commented):
Sorry for the nits...

DonIsaac committed:
feat(ast_tools): support `#[scope(exit_before)]` (#6350)

Closes #6311.

Adds support for `#[scope(exit_before)]`, which is the opposite of `#[scope(enter_before)]`
